walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Cases
+    // Options for no attendees - should be available all day.
+    if (request.getAttendees().size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Duration longer than a day will return empty array.
+    if (request.getDuration() > TimeRange.getTimeInMinutes(23, 59)) {
+      return Arrays.asList();
+    }
+    // Check if there are availabilities for all attendees
+    List<String> attendees = new ArrayList<>();
+    attendees.addAll(request.getAttendees());
+    attendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> allAttendeesAvailabilities = findAvailability(events, request, attendees);
+
+    // If there are no available timeslots for all attendees including optional attendees,
+    // Find availabilities for the mandatory attendees.
+    if (allAttendeesAvailabilities.size() == 0) {
+      return findAvailability(events, request, request.getAttendees());
+    }
+
+    return allAttendeesAvailabilities;
+  }
+
+  /** Private helper method to find all availabilities of people from the request. */
+  private Collection<TimeRange> findAvailability(
+      Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Find all the unavailable times of the attendees.
+    List<TimeRange> unavailabilities = new ArrayList<>();
+    for (Event event : events) {
+      if (attendees.stream().anyMatch(s -> event.getAttendees().contains(s))) {
+        unavailabilities.add(event.getWhen());
+      }
+    }
+    // If there are no unavailable times, return whole day.
+    if (unavailabilities.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    List<TimeRange> mergedUnavailabilities =
+        mergeAllUnavailabilities((ArrayList<TimeRange>) unavailabilities);
```

Comment:
Please avoid the cast.
If mergeAllUnavailabilities could work with any List<TimeRange>, it should accept any such List. Otherwise, _availabilities_ should be of the appropriate type.

Reply:
Done